Implement transformation to / from CoordinateFrame from / to related WCS

cmarmo opened the pull request:
### Description

This pull request Implement transformation to / from CoordinateFrame from / to related WCS.
This new frames are meant to be used for body-fixed reference frames for spheroidal Solar System bodies.
The implementation of the already available related WCS is also added.

Towards https://github.com/astropy/astropy/issues/11170.

Thanks for considering it.

This work is funded by the Europlanet 2024 Research Infrastructure (RI) Grant.

nstarman commented:
Hi @cmarmo, the actual PR looks pretty good, but I'm not sure I understand what it's meant to do... Is this a base class for use in custom frames outside of Astropy, or are you intending to refactor some existing frames in Astropy? If the former, I think this needs some examples demonstrating its usage; if the latter, this PR would be a good place to do the refactor 😄.

cmarmo commented:
> I'm not sure I understand what it's meant to do

It is meant to connect [planetary WCS](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2018EA000388) with a dedicated frame in astropy.
Still draft because I want to build the `GeodeticRepresentation` using the FITS keywords defined in the paper and I'm trying to figure out how to do it, as they are not in the WCS structure.
I was also preparing an example.

> or are you intending to refactor some existing frames in Astropy

which kind of refactoring do you have in mind?

nstarman commented:
> what kind of refactoring do you have in mind?

I just saw this was a base class. If it was a base class for any existing frames, then that would be some small refactoring. If it's for new stuff, then carry on :)

mhvk commented:
@nstarman - see #11170 for the larger picture. Thoughts on how to go about it very welcome -- we did the easy bits, which was to set up the option to create geodetic representations for other planets.

cmarmo commented:
@nstarman I find the documentation about [creating a custom frame](https://docs.astropy.org/en/latest/coordinates/frames.html) quite detailed and usable with the base class introduced by this PR. Then I've added an [example](https://astropy--14820.org.readthedocs.build/en/14820/generated/examples/coordinates/plot_mars-coordinate-frame.html#sphx-glr-generated-examples-coordinates-plot-mars-coordinate-frame-py) instead.
No transformations are defined for now, so the example is indeed minimal.

mhvk reviewed (commented):
@cmarmo - good to see the next step! I think we need to be a bit careful here not to mix up the representation (which you did already) with the frame. If I have coordinates on Mars, then the representation can be cartesian, spherical, geodetic or bodycentric, but the frame does not care about that. The only thing it needs to know is that it is of Mars, so that, if needed, it can be transformed to something relative to the solar system barycentre, etc. Basically, the frame should be like ITRS is for Earth, i.e., something suitable to add to the frame transformation graph. Of course, there should be different frames for the different planets, since they have different transformations to barycentric. (In princple, there might also be different frames for, say, different precession or orbital models, but that is definitely for another time!)

This has consequences both for the example plot in the new documentation (so good to add that, it helps focus discussion!) and for WCS.

For the example, there should be only one `MarsCoordinateFrame`. I don't know what the standard is exactly, but given that there are different spheroids, safest would seem to make the default `Cartesian` (as it is for Earth, via ITRS). And then one can initialize this with different geodetic and bodycentric representations, i.e.,the following would all work,
```
mars = MarsCoordinateFrame(x=1000*u.km, y=2000*u.km, z=3000*u.km)
mars2 = MarsCoordinateFrame(lon=10*u.deg, lat=20*u.deg, representation_type=MarsBestFitAeroid)
```

Indeed, note that even now, with the changes you have already made and with your definition of `MarsBestFitAeroid` in your example, one can do
```
mars = ITRS(x=1000*u.km, y=2000*u.km, z=3000*u.km)
mars.represent_as(MarsBestFitAeroid)  # or .represent_as("marsbestfitaeroid")
# <MarsBestFitAeroid (lon, lat, height) in (rad, rad, m)
#     (1.10714872, 0.93481117, 357668.0311648)>
mars2 = ITRS(lon=10*u.deg, lat=10*u.deg, representation_type=MarsBestFitAeroid)
mars2.cartesian
# <CartesianRepresentation (x, y, z) in m
#     (3293561.1722263, 580743.6972773, 583553.24056641)>
```

For WCS, this means that one should not instantiate a `BodyBaseCoordinateFrame`, but rather for each known object make a subclass, probably best on first use (i.e., cache the result), though perhaps it is fine to just define the 8 that we need in a `dict` (maybe replacing the one you already have).

One thing: if the default becomes `Cartesian`, there seems to be little point in having `BodyBaseCoordinateFrame` for now - it doesn't really do anything! One might as well add `obstime` attribute where needed.

Anyway, after this long comment, how about the following:
1. Make the documentation plot a separate PR that just subclasses `BaseCoordinateFrame` (and maybe includes the ITRS test; alternatively, rewrite that test to make a Mars frame as you do in the plot example - since those do not get tested, would be nice to have it!
2. Here, focus on WCS. I think creating custom classes may be the best way forward for now, but let me ping @mcara and @nden as WCS maintainers too.
3. Eventually, move class creation to coordinates, but then it would be good to actually link the frames to the transform graph (so, e.g., one could know where to point a telescope to observe Olympus Mons). Since we do have the machinery in place to calculate barycentric coordinates of the planets, this should not be very hard, but perhaps out of scope here.

Note that I'm not really sure whether or not to keep your `BodyBaseCoordinateFrame` - I mostly think in the end we'd like it to do more and I don't want to preclude that (also, it should probably be `BaseBodyCoordinateFrame`...). I'm also worried I'm just not thinking this through completely properly - in a way, these coordinates are like `EarthLocation`, which is very similar but not identical to ITRS. If we can avoid that can of worms...

cmarmo commented on astropy/coordinates/builtin_frames/bodyframes.py:
```diff
@@ -0,0 +1,83 @@
+
+    obstime = TimeAttribute(default=None)
+
+    _wrap_angle = 180 * u.deg  # for longitude
```
I have removed `wrap_angle` from the frame for now. Indeed sunpy defines a `wrap_angle` but I'm going to focus on the general implementation before.

cmarmo commented on examples/coordinates/plot_mars-coordinate-frame.py:
```diff
@@ -0,0 +1,168 @@
+
+    """
+
+    default_representation = MarsBestFitAeroid
```
Addressed in #14990.

cmarmo commented:
I am converting this PR back to draft as #14990 needs to go in first.

cmarmo commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +119,10 @@ def _wcs_to_celestial_frame_builtin(wcs):
                 obstime=wcs.wcs.dateobs or None,
             )
+        elif ("LN" in xcoord and "LT" in ycoord and "H" not in xcoord) or (
+            "LON" in xcoord and "LAT" in ycoord
```
No, it is not, I have removed all the 'LON'  'LAT' references.

cmarmo commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +137,21 @@ def _wcs_to_celestial_frame_builtin(wcs):
+            if "Bodycentric" in wcs.wcs.name:
+                baserepresentation = BaseBodycentricRepresentation
+            else:
+                baserepresentation = BaseGeodeticRepresentation
```
Indeed, but I need wcslib 8 (in progress) to retrieve the radii as I'm doing here
https://github.com/oudeis-io/astropy/blob/c6279c39e59f81efe7a84081cd28ef13e0367158/astropy/wcs/utils.py#L128-L138
I was planning to offer two PRs : a basic one (this one) and the one supporting wcslib 8.
Should I merge the two?

cmarmo replied:
This is addressed now.

cmarmo commented:
Mark Calabretta has just released wcslib8.1, this pr is meant to work when built against wcslib versions >= 8.
May I offer a PR updating the shipped version of wcslib for astropy 6? And then I will find a way to check the wcslib version in this pr....

mcara commented:
I'll make  a PR to update bundled WCSLIB first thing in the morning. You can rebase this PR after that.

mcara commented:
A PR to update WCSLIB was made here: https://github.com/astropy/astropy/pull/15035

Once it is approved, please rebase this PR.

Thanks!

cmarmo commented:
Ready for review if someone has time to check... thanks!

I have two known issues still needing to be addressed.
- when willing to address this [comment](https://github.com/astropy/astropy/pull/14820#discussion_r1242412052), I realized that Astropy `ITRS` introduces the coordinates `TLON` and `TLAT` for angular Earth coordinates. I cannot find where those two are defined, are they in the WCS standard?
- I'm still missing a way to check the version of wcslib when `ASTROPY_USE_SYSTEM_WCSLIB=1` and wcslib version < 8 in order to avoid errors during the build

mhvk commented:
@cmarmo - on the `wcslib` version, I think we will need to set a minimum version for the external `wcslib`. Since it will be quite a while before the next astropy version is released, that is not a problem for users.

But I guess the bigger question is what to do in the meantime, since without some solution weekly cron would start to fail for the less usual architectures; I think in a previous case we ensured we had backports among our `sources.list` and used that to get an updated version of `libwcs` that @olebole had put there. Hopefully something like that will be possible again; @olebole, do you remember what we did before?

olebole commented:
@mhvk I created a special repository, [Astro backports](https://launchpad.net/~olebole/+archive/ubuntu/astro) where I have put a wcslib backport at that time  (that was wcslib-7.7 for Ubuntu 20.04). It is perfectly doable again. Just tell me, which Ubuntu version(s) is needed.

mhvk commented:
Super! So, the `ci_cron_weekly` runs still use ubuntu-20.04, so that should still work!  Though actually I am a bit confused why both ubuntu-20.04 and ubuntu-latest are shown... Anyway, I guess we would just have to add back the repository somewhere on https://github.com/astropy/astropy/blob/1ef68aa829db94b1c506afbb7f6fc2faeaf4eee5/.github/workflows/ci_cron_weekly.yml#L132-L145, presumably by adding to `sources.list` or so (if you have any links to where this was done before, let me know; I couldn't find it easily, sadly).

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +142,35 @@ def _wcs_to_celestial_frame_builtin(wcs):
+            if Version(_wcs.__version__) >= Version("8.0"):
+                a_radius = wcs.wcs.aux.a_radius
+                b_radius = wcs.wcs.aux.b_radius
+                c_radius = wcs.wcs.aux.c_radius
```
If we need the `if`, then there should be an `else` that raises an error that the wcslib version does not support this.

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -163,6 +216,28 @@ def _celestial_frame_to_wcs_builtin(frame, projection="TAN"):
+            wcs.wcs.name = "Planetographic Body-Fixed"
+        elif issubclass(frame.representation_type, BaseBodycentricRepresentation):
+            wcs.wcs.name = "Bodycentric Body-Fixed"
+        if Version(_wcs.__version__) >= Version("8.0"):
```
If we do need a version check, then here it would be fine to do `if WCSLIB_LT_8: raise ...`

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -163,6 +216,28 @@ def _celestial_frame_to_wcs_builtin(frame, projection="TAN"):
+                wcs.wcs.aux.b_radius = (
+                    frame.representation_type._equatorial_radius.value
+                )
+            if hasattr(frame.representation_type, "_flattening"):
```
Again, no `if` statement necessary - when we get here, the representation should have `_flattening`

mhvk reviewed (commented):
This looks nice! My main comment is that we need somehow to have a different representation any time we encounter a different equatorial radius or flattening. We probably should make a cache for those too...

mcara commented on astropy/wcs/utils.py:
```diff
@@ -31,6 +39,22 @@
+}
+
+
+solar_system_body_frames = {}
```
Why is this a global variable?

cmarmo replied:
Sorry... fixed.

mcara commented on astropy/wcs/utils.py:
```diff
@@ -31,6 +39,22 @@
 ]
 
 
+solar_system_obj_dict = {
```
I believe global "constants"'s names should be capitalized, see https://peps.python.org/pep-0008/#constants

mcara commented:
> * I'm still missing a way to check the version of wcslib when `ASTROPY_USE_SYSTEM_WCSLIB=1` and wcslib version < 8 in order to avoid errors during the build

Doesn't `_wcs.__version__` work with system installed WCSLIB?

mhvk commented:
@mcara - the problem is that with the old `libwcs7`, one cannot *build* the changes that @cmarmo is introducing here, as it relies on new items on structs.

mcara commented:
Ah, yes. What about using `#if WCSLIB_VERSION...#endif` directives to "turn off" new code at compile time? It should work, I believe. However, I do not think we ever did anything like this in the past (disable parts of the WCSLIB wrapper code depending on the WCSLIB version).

mhvk commented:
My 2¢ is that we should avoid these issues by `libwcs8` being a build requirement... I think this is really relevant only for distributions which can ensure the latest version is packaged; for developers, one has to do work to avoid using the built-in version.

cmarmo commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +142,35 @@ def _wcs_to_celestial_frame_builtin(wcs):
+                b_radius = wcs.wcs.aux.b_radius
+                c_radius = wcs.wcs.aux.c_radius
+
+            if "bodycentric" in wcs.wcs.name.lower():
```
There is no standard way to switch from geodetic to bodycentric representations in FITS or  wcslib. We did not want to introduce new keywords for that, so in https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2018EA000388 we suggest to use `WCSNAME` for that.

mhvk replied:
OK, that makes sense. Probably useful for future reference to give that link in a comment, something like:
```
# Coordinates on a planetary body, as defined in
# https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2018EA000388
```

cmarmo commented on astropy/wcs/utils.py:
```diff
@@ -163,6 +198,19 @@ def _celestial_frame_to_wcs_builtin(frame, projection="TAN"):
+            ycoord = "LAT-"
+        if frame.obstime is not None:
+            wcs.wcs.dateobs = frame.obstime.utc.isot
+        if issubclass(frame.representation_type, BaseGeodeticRepresentation):
```
You mean ITRS, right?
I am confused about the ITRS WCS definition, because of the `TLON` `TLAT` descriptions which I cannot find in the literature... if someone has some references this will help me a lot... Thanks!

mhvk replied:
Indeed, I meant ITRS. I don't know where `TLON`, `TLAT` comes from either... But google gives, e.g., http://tdc-www.harvard.edu/software/wcstools/wcstools.multiwcs.html and https://www.astro.rug.nl/software/kapteyn/allsky.html, so I guess we should just continue to support it.

cmarmo replied:
Thanks! I coudn't find those links.

cmarmo commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -323,6 +338,38 @@ def test_wcs_to_celestial_frame():
 
+    mywcs = WCS(naxis=2)
+    mywcs.wcs.ctype = ["MALN-TAN", "MALT-TAN"]
+    mywcs.wcs.radesys = "ICRS"
```
Indeed, this depends on obstime (see the section *2.2.1 FITS WCS Keywords and Their Recommended Use in Planetary Surface Context* in https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2018EA000388) ... I have oversimplified for now ...

cmarmo commented:
I think I have addressed most of the comments.
I resolved the related comments as this helps me to go through the issue, but if you are not happy with the solution feel free to unresolve.

> What about using `#if WCSLIB_VERSION...#endif` directives to "turn off" new code at compile time? It should work, I believe. However, I do not think we ever did anything like this in the past (disable parts of the WCSLIB wrapper code depending on the WCSLIB version).

I tried that, but I had issues with its "floating point" representation ... all my effort to define a macro using predefined info at the time of the compilation failed ... so I'm glad if this is not required ... :innocent:

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -353,6 +370,59 @@ def test_wcs_to_celestial_frame():
+        triaxial_wcs.wcs.aux.a_radius = 2439700.0
+        triaxial_wcs.wcs.aux.b_radius = 2439900.0
+        triaxial_wcs.wcs.aux.c_radius = 2438800.0
+    with pytest.raises(
```
This should fail on `WCSLIB_LT_8` - which I guess we are not currently testing...

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -353,6 +370,59 @@ def test_wcs_to_celestial_frame():
+
+    mywcs = WCS(naxis=2)
+    mywcs.wcs.ctype = ["MALN-TAN", "MALT-TAN"]
+    mywcs.wcs.radesys = "ICRS"
```
I'm still confused why it is necessary to set this.

cmarmo replied:
Removed them: it is not necessary, indeed.

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -473,6 +543,44 @@ class FakeFrame(BaseCoordinateFrame):
+    with pytest.raises(
+        ValueError, match="The representation type should be geodetic or bodycentric"
+    ):
+        mywcs = celestial_frame_to_wcs(frame, projection="CAR")
```
No need to assign it to anything (i.e., delete "mywcs = ")

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -30,6 +38,20 @@
     "fit_wcs_from_points",
 ]
 
+WCSLIB_LT8 = Version(_wcs.__version__) < Version("8.0")
```
Our nomenclature has been to use underscores on both sides of LT, i.e., `WCSLIB_LT_8`

cmarmo replied:
... I checked here
https://github.com/astropy/astropy/blob/98b399d52d730d91c2cd29babe1e8545caff219c/astropy/wcs/wcs.py#L106-L108
That's why I didn't put the underscore ... is it the wrong place?

mhvk replied:
OK, no problem! I'm just used to how we do it with numpy...

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +143,76 @@ def _wcs_to_celestial_frame_builtin(wcs):
+
+            if WCSLIB_LT8:
+                warnings.warn(
+                    f" {_wcs.__version__} does not support planetary WCS,"
```
If we need 8.1, then that should be the version check too, i.e., `WCSLIB_LT_8_1`

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +143,76 @@ def _wcs_to_celestial_frame_builtin(wcs):
+                        "triaxial systems are not supported at this time."
+                    )
+
+            # instance a new representation class
```
instance -> create

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +142,35 @@ def _wcs_to_celestial_frame_builtin(wcs):
+
+            frame = solar_system_body_frames.get(xcoord[:2])
+            object_name = solar_system_obj_dict.get(xcoord[:2])
+            if frame is None and object_name is not None:
```
`object_name` can no longer be `None`, correct? So, remove that part of the `if` statement.

cmarmo replied:
Right ... sorry...

mhvk reviewed (commented):
I think this is basically ready, except that we should set up `libwcs8` for our tests with system libraries. @olebole - can you indeed upload it to that previous archive? And then we'll have to work out how to use that again...

mhvk commented:
p.s. If we do have `libwcs8` set up, all those `if WCSLIB_LT_8` statements can go...

cmarmo commented:
> p.s. If we do have `libwcs8` set up, all those `if WCSLIB_LT_8` statements can go...

Thanks for clarifying... I think I have addressed the comments, but I don't understand how to test the condition on the version ...

olebole commented:
Just to make sure: do you need this for 20.04 (focal) or 22.04 (jammy)?

mhvk commented:
@olebole - simple question but not so easy to find out! We run these tests on other architectures, in https://github.com/astropy/astropy/blob/79dd50f61ef5e0a568dd42549319db38fefd0bc7/.github/workflows/ci_cron_weekly.yml#L96-L155
This uses https://github.com/uraimo/run-on-arch-action which supports all kinds of distributions. From our entries, we use `ubuntu_rolling`, and if I look at where files are actually downloaded from, it looks like it is `lunar` (see https://github.com/astropy/astropy/actions/runs/5548242376/job/15029068040#step:3:54).

That is recent enough that I would think we  might just be able to get the package you will put on debian testing anyway!?

olebole commented:
Lunar is Ubuntu 23.04, and this was before I uploaded wcslib-8. In the next version (23.10) it will be there.
I now uploaded one for 20.04 and one for 22.04; I can add one for 23.04 without problems.

mhvk commented:
Thanks! What I still need to find out is how we dealt with the extra `apt` source... If you happen to remember, let me know!

olebole commented:
I now uploaded wcslib-8.1 for Focal (20.04), Jammy (22.04), and Lunar (23.04). Later Ubuntu releases (from Mantic/23.10) will have wcslib-8.1 in the distribution.
For the use:
```
add-apt-repository ppa:astropy/test-support
apt update
```
The web view of the repository is https://launchpad.net/~astropy/+archive/ubuntu/test-support
It will still take a few hours until all builds are complete and published there.

mhvk commented:
Hmm, I cannot seem to get this to work, /perhaps/ because of a missing certificate, though I also wonder if we don't need `libwcs8` as well as `wcslib-dev`... See #15083 for my trials. I started with `add-apt-repository ppa:astropy/test-support` but that command is not available (and I didn't seem able to install `software-properties-common` either...). Then I tried more simply,
```
            # Add test-support repository for wcslib8
            echo "deb https://ppa.launchpadcontent.net/astropy/test-support/ubuntu lunar main" > /etc/apt/sources.list.d/test-support.list
```
but this gives
```
 Err:2 https://ppa.launchpadcontent.net/astropy/test-support/ubuntu lunar InRelease
    Certificate verification failed: The certificate is NOT trusted. The certificate issuer is unknown.  Could not handshake: Error in the certificate verification. [IP: 185.125.190.52 443]
```
I tried installing `ca-certificates` before this, and adding `[trusted=yes]`, but neither helped (see the various commits in #15083). Any suggestions?

olebole commented:
The certificate error is for the SSL connection; you could use **http://** instead of **https://**. The security is till there, as the packages are signed with my GPG key (and this should be in the keyring).

cmarmo commented:
Hello, assuming that #15083 is the way to fix the weekly builds, am I right removing `WCSLIB_LT8` everywhere? I cannot find a use case where it is still needed...

cmarmo commented:
Hi @mhvk , all is green: I have added the wcslib8 repo from https://github.com/astropy/astropy/pull/15083 and removed the tests on wcslib version.
Do you think this is mergeable now?
Thanks a lot!

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -353,6 +368,52 @@ def test_wcs_to_celestial_frame():
 
 
+def test_wcs_to_body_frame():
+    # Import astropy.coordinates here to avoid circular imports
```
I don't see how you can get a circular import in a test function; just put it on top!

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -353,6 +368,52 @@ def test_wcs_to_celestial_frame():
+    assert issubclass(frame, BaseCoordinateFrame)
+    assert issubclass(frame.representation_type, BaseBodycentricRepresentation)
+    assert frame.name == "Mars"
+    assert frame.representation_type._equatorial_radius == 3396190.0 * u.m
```
Also test `_flattening == 0.0`?

cmarmo replied:
I have tested Mars flattening and added a case where flattening == 0 (Venus)

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -353,6 +368,52 @@ def test_wcs_to_celestial_frame():
+    frame = wcs_to_celestial_frame(mywcs)
+    assert issubclass(frame, BaseCoordinateFrame)
+    assert issubclass(frame.representation_type, BaseGeodeticRepresentation)
+    assert frame.representation_type._equatorial_radius == 6378137.0 * u.m
```
Maybe also test `_flattening`?

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -473,6 +534,44 @@ class FakeFrame(BaseCoordinateFrame):
 
 
+def test_body_to_wcs_frame():
+    # Import astropy.coordinates here to avoid circular imports
```
Again, don't see how that can be a problem!

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -473,6 +534,44 @@ class FakeFrame(BaseCoordinateFrame):
+        name = "Mars"
+        representation_type = SphericalRepresentation
+
+    frame = IAUMARSSphereFrame()
```
Maybe again do the failures at the end?

cmarmo replied:
not sure my understanding is correct here: I have moved the test on the error after the definitions of the classes.

mhvk replied:
No, that is not what I meant: I just meant that in this function I feel it is better first to test things as they are supposed to work, and then at the end test that the right errors are raised. I.e., move this `pytest.raises` to the very end of the test function.

cmarmo replied:
Understood, sorry.

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -473,6 +534,44 @@ class FakeFrame(BaseCoordinateFrame):
+    mywcs = celestial_frame_to_wcs(frame, projection="CAR")
+    assert mywcs.wcs.ctype[0] == "MALN-CAR"
+    assert mywcs.wcs.ctype[1] == "MALT-CAR"
+    assert mywcs.wcs.name == "Planetographic Body-Fixed"
```
I'd test abc_radius for this one too.

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +140,60 @@ def _wcs_to_celestial_frame_builtin(wcs):
+                )
+
+            # create a new representation class
+            representation_type = solar_system_body_representation_type.get(xcoord[:2])
```
If we keep a cache (which I think we should), we have to be sure it is correct. So, we need a new instance any time a new object *or* a different representation *or* new equatorial/radius are used. That means we should set the key to `object_name, baserepresentation, equatorial_radius, flattening`.

Thinking more about this, I think I would implement it with `lru_cache` from [functools](https://docs.python.org/3/library/functools.html), as follows:
```
from functools import lru_cache

@lru_cache(maxsize=100)
def solar_system_body_representation_type(object_name, baserepresentation, equatorial_radius, flattening):
    return type(
        f"{object_name}{baserepresentation.__name__[4:]},
        (baserepresentations,),
        dict(_equatorial_radius=equatorial_radius, _flattening=flattening)
    )
```

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +140,60 @@ def _wcs_to_celestial_frame_builtin(wcs):
+                solar_system_body_representation_type[xcoord[:2]] = representation_type
+
+            # create a new frame class
+            frame = solar_system_body_frames.get(xcoord[:2])
```
Same as above: this should be a different frame any time there is a new `object_name` *or* `representation_type`, so again something like
```
@lru_cache(maxsize=100)
def solar_system_body_frame(object_name, representation_type):
    return type(
        f"{object_name}Frame",
        (BaseCoordinateFrame,),
        dict(...),
    )
```

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -72,6 +91,9 @@ def _wcs_to_celestial_frame_builtin(wcs):
     # Import astropy.time here otherwise setup.py fails before extensions are compiled
     from astropy.time import Time
 
+    solar_system_body_frames = {}
```
We want these to be global variables, so that we define the types only once, so they should be outside of the function.

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +140,60 @@ def _wcs_to_celestial_frame_builtin(wcs):
                 obstime=wcs.wcs.dateobs or None,
             )
+        elif (
+            "LN" in xcoord
```
Could this be more strict, i.e., `xcoord[2:4] in ("LN", "LT") and "H" not in xcoord and "CR" not in xcoord`?

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -163,6 +239,30 @@ def _celestial_frame_to_wcs_builtin(frame, projection="TAN"):
         wcs.wcs.radesys = "ITRS"
         wcs.wcs.dateobs = frame.obstime.utc.isot
+        if issubclass(frame.representation_type, BaseGeodeticRepresentation):
+            wcs.wcs.name = "Geodetic Terrestrial Representation"
```
Is it tested that these round-trip? It would be nice to have it for sure! (Note I'm also OK with just not touching this part, but if we add anything we should test it...)

cmarmo replied:
As you are OK with just not touching them and I feel uncomfortable in adding ad hoc wcs descriptions about reference systems out of the scope of this PR, I'd rather remove this lines and the following.

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -163,6 +239,30 @@ def _celestial_frame_to_wcs_builtin(frame, projection="TAN"):
+        if issubclass(frame.representation_type, BaseGeodeticRepresentation):
+            wcs.wcs.name = "Geodetic Terrestrial Representation"
+        elif issubclass(frame.representation_type, BaseBodycentricRepresentation):
+            wcs.wcs.name = "Angular Geocentric Representation"
```
Does this indeed round-trip?

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -163,6 +239,30 @@ def _celestial_frame_to_wcs_builtin(frame, projection="TAN"):
+            wcs.wcs.name = "Bodycentric Body-Fixed"
+        else:
+            raise ValueError(
+                "The representation type should be geodetic or bodycentric,"
```
Maybe rephrase by turning around "planetary coordinates in WCS require a geodetic or bodycentric representation, not {frame.representation_type}."

mhvk reviewed (commented):
@cmarmo - thanks for the reminder; this has been a while, sorry!

I think it looks very good but apart from the nitpicks we're now getting too, two comments:
1. I think we should cache representations and frames that were used before, and thinking more about it, believe using a function with `@lru_cache` would be best; see in-line comments.
2. Are we in a state where we could also add a round-trip test? I.e., make a `WCS` from a frame and read it back in, to see that it is the same? It may be that this does not work directly, since the frames are not identical... But the reverse, make a frame from a WCS and write the result to a new WCS, and check that those are the same, should work I think (may need to choose `abc_radius` such that `_flattening` is exact, say `2**-4` or so).

pep8speaks commented:
Hello @cmarmo! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`astropy/wcs/tests/test_utils.py`](https://github.com/astropy/astropy/blob/eeaac9f52338979d5f8b4a0a1e6c9d4f59c15c7c/astropy/wcs/tests/test_utils.py):

> [Line 562:89](https://github.com/astropy/astropy/blob/eeaac9f52338979d5f8b4a0a1e6c9d4f59c15c7c/astropy/wcs/tests/test_utils.py#L562): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 88 characters)

mhvk commented:
Great! Do ping when all are done, and I'll re-review.

cmarmo commented:
@mhvk I believe I have addressed your comments.
Thanks for your patience!

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +158,62 @@ def _wcs_to_celestial_frame_builtin(wcs):
+                equatorial_radius,
+                flattening,
+            )
+            if representation_type is None:
```
This whole clause is no longer needed - the above function call will either create the representation or return the cached function!

Indeed, this is why there is no coverage of these lines.

mhvk commented on astropy/wcs/utils.py:
```diff
@@ -118,6 +158,62 @@ def _wcs_to_celestial_frame_builtin(wcs):
+            frame = solar_system_body_frame(
+                SOLAR_SYSTEM_OBJ_DICT.get(xcoord[:2]), representation_type
+            )
+            if frame is None:
```
Same, this is no longer needed - the function call will do it.

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -353,6 +368,66 @@ def test_wcs_to_celestial_frame():
 
 
+def test_wcs_to_body_frame():
+    unknown_wcs = WCS(naxis=2)
```
Like for the other direction, I would move the tests for failures to the very end of the test function. I think it is more helpful if a test function first shows how things are supposed to work, and then checks that you get an appropriate error if you do something incorrectly.

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -353,6 +368,66 @@ def test_wcs_to_celestial_frame():
+    ):
+        frame = wcs_to_celestial_frame(triaxial_wcs)
+
+    mywcs = WCS(naxis=2)
```
So, I would start the test function with this part.

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -353,6 +368,66 @@ def test_wcs_to_celestial_frame():
+    assert frame.name == "Venus"
+    assert frame.representation_type._equatorial_radius == 6051800.0 * u.m
+    assert frame.representation_type._flattening == 0.0
+
```
Could you add a test that the same frame is returned if you do this twice? I.e.,
```
# Check that frames are cached appropriately
frame2 = wcs_to_celestial_frame(mywcs)
assert frame2.representation_type is frame.representation_type
assert  frame2 is frame
```

mhvk commented on astropy/wcs/tests/test_utils.py:
```diff
@@ -353,6 +368,66 @@ def test_wcs_to_celestial_frame():
+    mywcs.wcs.aux.a_radius = 3396190.0
+    mywcs.wcs.aux.b_radius = 3396190.0
+    mywcs.wcs.aux.c_radius = 3376190.0
+    frame = wcs_to_celestial_frame(mywcs)
```
Maybe call this `frame3`, so that one has
```
frame3 = wcs_to_celestial_frame(mywcs)
assert frame3.representation_type is not frame.representation_type
assert frame3 is not frame
```

mhvk reviewed (commented):
Thanks! This seems almost done: the main code now has two unnecessary clauses, and I clarified what I meant with reorganizing the tests.

cmarmo commented:
Better now? :grin:

mhvk reviewed (approved):
Looks all great now!  Let's get it in!

mhvk commented:
@cmarmo  - there's a bit of a what's-new already, but I realize now (after merging, of course, sigh) that it may be good to add a little to it, noting that WCS can also be written (with perhaps a link to the article describing the new standard). Perhaps add to the documentation too, if you think that is already useful.

cmarmo commented:
> @cmarmo - there's a bit of a what's-new already, but I realize now (after merging, of course, sigh) that it may be good to add a little to it, noting that WCS can also be written (with perhaps a link to the article describing the new standard). Perhaps add to the documentation too, if you think that is already useful.

PR coming soon.
Thanks a lot for your time and your patience @mhvk !

pllim commented:
Does this have anything to do with this pyinstaller failure I started seeing?

https://github.com/astropy/astropy/actions/runs/6294343966/job/17086053141

mhvk commented:
@pllim - I fear so. I think the combination of importing and `setup/teardown` just become too much for pytest to handle. Since really it does not matter exactly what is used in the tests, I just duplicated some definitions in #15381. Hopefully, with that, things will run (I used extra CI since I assume that's why we didn't notice the failures before).
